### PR TITLE
Converting many-client to async

### PR DESCRIPTION
**Pull request by Denommus**
## Description

This PR  converts ManyClient to be an async client, which is useful when a user wants to utilize many-client as a library.

Also adds a `LedgerClient` for specific requests to the ledger server. More clients are coming as tests are created for them.

## Related Issue

Fixes many-protocol/specification#55

## Testing

The tests cover the current changes on ManyClient. LedgerClient is being tested in [many-protocol/specification](https://github.com/many-protocol/specification) gherkin tests.

## Breaking Changes (if applicable)

Since this is async, newer clients need to take async into consideration and call `.await` in many of the methods. This is made obvious with Rust's type system.

## Checklist:

- [X] I have read and followed the CONTRIBUTING guidelines for this project.
- [X] I have added or updated tests and they pass.
- [X] I have added or updated documentation and it is accurate.
- [X] I have noted any breaking changes in this module or downstream modules.

### CODE REVIEW

File: src/many-client/src/client/ledger.rs
Comment (hansl) [inline src/many-client/src/client/ledger.rs:21]: The functions exposed by this structure should match the specification, what do you think? So `info()`, `balance()`, `send()`, ...

Comment (Denommus) [reply src/many-client/src/client/ledger.rs:21]: I agree, but I'm thinking of adding more methods as I write specification tests for them.

**Review by fmorency — COMMENTED** [review]
So far, so good, thanks! Is this ready to merge, or are you still adding new clients?

**Comment by Denommus** [comment]
@fmorency I'll add more methods to LedgerClient and then add more clients in future PRs.

File: src/many-client/src/client/ledger.rs
Comment (Denommus) [reply src/many-client/src/client/ledger.rs:21]: @hansl I thought more about it and I've matched those, I'll adapt the specification tests to match

File: src/many-types/src/lib.rs
Comment (hansl) [inline src/many-types/src/lib.rs:49]: Note: I wouldn't mind an enum. There never was a goal to have idempotence between encoding and decoding.

**Review by hansl — APPROVED** [review]
LGTM, but please wait for Felix's review before merging (and if you add more commits please request my review again). Thanks!

File: src/many-client/src/client/ledger.rs
Comment (fmorency) [inline src/many-client/src/client/ledger.rs:49]: I don't believe this is needed if the original `ManyClient` is properly initialized.

**Review by fmorency — CHANGES_REQUESTED** [review]

File: src/many-client/src/client/ledger.rs
Comment (Denommus) [reply src/many-client/src/client/ledger.rs:49]: @fmorency the idea here is that on the specification tests I have a single ledger client initialized with the faucet private key, but I still want to send tokens from another private key. Do you recommend me to have one ledger client per private key instead, then?

Comment (fmorency) [reply src/many-client/src/client/ledger.rs]: This is related to https://github.com/many-protocol/specification/pull/53#discussion_r954985507

If `to` is anonymous, then "the first server who receives this message should assume he is the destination."

See https://github.com/many-protocol/specification/blob/6ba25eebec3493340e6537682eb360ba24046042/spec/cddl/many.cddl#L12-L16

Comment (fmorency) [reply src/many-client/src/client/ledger.rs]: I am suggesting to defer this configuration where the `ManyClient` is initialized.

File: src/many-client-macros/src/lib.rs
Comment (fmorency) [inline src/many-client-macros/src/lib.rs:73]: rm

File: src/many-client-macros/src/lib.rs
Comment (fmorency) [inline src/many-client-macros/src/lib.rs:101]: I would rephrase this to: "The base struct should be a unit struct", or something like that, since `struct Foo;` are known as unit structs.

**Review by fmorency — APPROVED** [review]
LGTM! Thanks, @Denommus.

Approving assuming comments are addressed.

File: src/many-client/src/client/ledger.rs
Comment (fmorency) [reply src/many-client/src/client/ledger.rs]: Your recent changes fix this.

File: src/many-client/src/client/ledger.rs
Comment (hansl) [inline src/many-client/src/client/ledger.rs:5]: It's a great start, but I found the way to declare methods confusing and easy to mistake. I would have preferred something that's a bit more rust-like for declaring the API. Something like:

```rust
#[many_client]
pub trait LedgerClientTrait {
  fn info(&self) -> Result<InfoReturns, ManyError>;
  fn balance(&self, args: BalanceArgs) -> Result<BalanceReturns, ManyError>;
  fn send(&self, args: SendArgs) -> Result<(), ManyError>;
}

pub struct LedgerClient(ManyClient);

impl LedgerClient {
  pub fn new(client: ManyClient) -> Self {
    Self(client)
  }
}

#[many_client]
impl LedgerClientTrait for LedgerClient {}
```

Or something similar. What I really want is the `fn balance(&self, args: BalanceArgs) -> Result<BalanceReturns, ManyError>` part which is a bit more code, but less confusing to readers and IDEs.

What do you think?

File: src/many-types/src/lib.rs
Comment (Denommus) [reply src/many-types/src/lib.rs:49]: @hansl I think the internal representation is fine, I just added that comment for clarity on what the abstraction actually means.

File: src/many-client/src/client/ledger.rs
Comment (Denommus) [reply src/many-client/src/client/ledger.rs:5]: @hansl I like the idea. I think the first `many_client` in that block is unnecessary, but otherwise it's seems like a nice interface.

Comment (Denommus) [reply src/many-client/src/client/ledger.rs:5]: @hansl I like the idea of that proposal, but I think the interface would look something more like this:

```rust
#[many_client(LedgerClient)]
trait AnyTraitName {
  fn info(&self) -> Result<InfoReturns, ManyError>;
...
}

struct LedgerClient(ManyClient);
```

And then it expands to a normal impl (instead of a trait impl)

File: src/many-client-macros/src/lib.rs
Comment (hansl) [inline src/many-client-macros/src/lib.rs:14]: I wonder if you can't just `parse2::<(Type, Option<LitStr>)>` or similar. I know you can declare your own type and `impl syn::parse::Parse for ...`. Look at how I did it in `many-macros/src/lib.rs` for the endpoint which is what you want.

There's a bug here where I could do `many_client("hello", "world", SomeStruct, "blue")`. It's not proper.

**Review by hansl — CHANGES_REQUESTED** [review]
This is getting really good. Just a few comments about the macro code itself.

**Review by fmorency — APPROVED** [review]
I'll let @hansl review the macro in-depth, but from my perspective, everything LGTM. Thanks!

**Review by hansl — APPROVED** [review]
LGTM, we can iterate over this macro later too.

**Comment by hansl** [comment]
uh... I think CI was failing? This should not have auto merged.

**Pull request merged**
